### PR TITLE
WordPressComponentProps: Only add refs when they are actually added

### PR DESCRIPTION
**Pull request by mirka**
## What?

Remove the `ref` prop from `WordPressComponentProps`.

The ad hoc `Omit`s on the following components have been removed:

- CheckboxControl
- Disabled
- FormToggle
- Placeholder
- RadioControl
- TextareaControl

## Why?

The `WordPressComponentProps` type always added a `ref` prop, even though the `props` object of a `ForwardRefRenderFunction` (i.e. the function to be passed to `forwardRef()`) should never even contain a `ref` prop.

The `ref` prop should only be mixed into the `props` object type only after the component is wrapped by `forwardRef()`.

## How?

Don't include the `ref` in `WordPressComponentProps`, but instead allow it to be added from `WordPressComponentFromProps` or `WordPressComponent`.

## Testing Instructions

1. `npm run storybook:dev`
2. In the Docs view for the affected components listed above, see that the `ref` prop is still omitted from the props table.

### CODE REVIEW

File: packages/components/src/ui/context/wordpress-component.ts
Comment (mirka) [inline packages/components/src/ui/context/wordpress-component.ts:12]: Moved these from JSDoc to normal comments because they're more about implementation details and not useful in IntelliSense.

**Review by ciampo — COMMENTED** [review]
Code changes look good and the project builds correctly, including Storybook.

> I don't love that `WordPressComponentProps` now has four parameters, but in the end it seemed easier to maintain than creating a separate `WordPressComponentPropsWithoutRef` type.

I'm also not in love with the 4th parameter — it's not easy to discover, and it forces devs to define the previous parameters (even if they otherwise could have been omitted in favor of their default value).

Adding a `WordPressComponentPropsWithoutRef` would also align better with the fact that React exposes a `ComponentPropsWithRef` and a `ComponentPropsWithoutRef`.

How much more complicated would the implementation be, if we had to go that way?

**Comment by mirka** [comment]
Thanks for nudging me to try harder, I think I uncovered a critical detail that was straight-up wrong. As a result, it's much more simpler now.

I've updated the PR description, and will add a few inline comments.

File: packages/components/src/ui/context/test/wordpress-component.tsx
Comment (mirka) [inline packages/components/src/ui/context/test/wordpress-component.tsx:24]: This code would not trigger a TS error on trunk, which is incorrect.

File: packages/components/src/ui/context/test/wordpress-component.tsx
Comment (mirka) [inline packages/components/src/ui/context/test/wordpress-component.tsx:16]: What do you think about test files for static checks?? 😆 It adds some unnecessary time to the jest queue but it feels useful to have well-documented "tests" for our TS types as they get more complex. I guess we could also just put the code in a file without wrapping it in `describe`/`it` calls and add comments. Not sure which is more understandable to a person who stumbles on the file for the first time.

File: packages/components/src/ui/context/wordpress-component.ts
Comment (mirka) [inline packages/components/src/ui/context/wordpress-component.ts:18]:
```diff
- React.ComponentPropsWithRef
+ React.ComponentPropsWithoutRef
```

This here is the critical change.

File: packages/components/src/ui/context/wordpress-component.ts
Comment (mirka) [inline packages/components/src/ui/context/wordpress-component.ts:52]: The `ForwardsRef` boolean is now here. But I'm going to give #43611 another try and see if I can streamline it more.

File: packages/components/src/view/component.tsx
Comment (mirka) [inline packages/components/src/view/component.tsx:37]: Here we explicitly add the `ref` because `styled.div` accepts it.

(This part is kind of hacky to begin with 🙁 I might be able to untangle it if I understood what the `selector` property is for...)

File: packages/components/src/view/stories/index.tsx
Comment (mirka) [inline packages/components/src/view/stories/index.tsx:27]: I just added a quick story so I could verify in Docs view that I didn't mess up the props.

File: packages/components/src/ui/context/test/wordpress-component.tsx
Comment (ciampo) [reply packages/components/src/ui/context/test/wordpress-component.tsx:16]: The current approach sounds reasonable to me! Let's keep it as-is and iterate if necessary

File: packages/components/src/view/component.tsx
Comment (ciampo) [reply packages/components/src/view/component.tsx:37]: > I might be able to untangle it 

What do you have in mind?

> if I understood what the `selector` property is for...

I believe it's used in `contextConnect`:

https://github.com/WordPress/gutenberg/blob/d64dab9645c1ede4e53b36ad4f96637b1766b25e/packages/components/src/ui/context/context-connect.ts#L76-L77

Explanation:

https://github.com/WordPress/gutenberg/blob/515928d0ae52aa9fb767a6a51d31e7050d07ad73/packages/components/src/ui/context/wordpress-component.ts#L39-L47

File: packages/components/src/view/stories/index.tsx
Comment (ciampo) [reply packages/components/src/view/stories/index.tsx:27]: Do you think it's worth keeping this story ?

**Review by ciampo — COMMENTED** [review]
I also think that things look much better!

Just a few questions left:
 - Should we wait for any more progress happening in #43611, before merging this PR?
 - Are there any components that already accept `ref` as a normal prop (i.e. without being explicitly wrapped in `forwardRef`)? If so, is the plan for them to refactor them gradually as we convert them to TypeScript, or should we refactor them as part of this PR?
 - Do we need to make any changes to the contributing guidelines, especially re. handling `ref`s as part of the TypeScript migration ?

File: packages/components/src/view/component.tsx
Comment (mirka) [reply packages/components/src/view/component.tsx:37]: > component interpolation in styled components

Does this refer to this kind of thing? 👇

```js
styled.div`
  ${ MyComponent } {
    color: black;
  }
`;
```

>> I might be able to untangle it

> What do you have in mind?

A lot of the `ts-ignore`s have to do with the fact that the `WordPressComponent` type defines the custom properties `select` and `__contextSystemKey__` as required. Things could be easier if they were marked as optional.

Another thing is this line:

https://github.com/WordPress/gutenberg/blob/5a63b2950eeab3ee09afa00a1f626cd2c1bfbaf7/packages/components/src/ui/context/context-connect.ts#L55

I don't see how `WrappedComponent` could ever have a `__contextSystemKey__` property at that point. If someone ran the same component through multiple `contextConnect`s? So either I don't fully understand what's going on, or this "merged namespaces" stuff can be cut.

---

In any case, these concerns are in follow-up territory.

File: packages/components/src/view/stories/index.tsx
Comment (mirka) [reply packages/components/src/view/stories/index.tsx:27]: No strong opinion. But `View` is actually a publicly exported component (albeit experimental), so I think it makes sense for it to have a place in Storybook, if only to see what it is.

**Comment by mirka** [comment]
> Should we wait for any more progress happening in [#43611](https://github.com/WordPress/gutenberg/pull/43611), before merging this PR?

Nope, we're at a good point to merge 👍

> Are there any components that already accept ref as a normal prop (i.e. without being explicitly wrapped in forwardRef)? If so, is the plan for them to refactor them gradually as we convert them to TypeScript, or should we refactor them as part of this PR?

Good question. The good thing about the approach in this PR is that it immediately corrects these mistakes, with no refactoring required.

Sweeping the components package, I found these components to accept `ref` props when they shouldn't:

- BaseControl.VisualLabel
- InputControl ▸ Label
- RangeControl ▸ RangeMark, RangeRail, Marks, SimpleTooltip
- FormGroup ▸ FormGroupContent, FormGroupLabel
- UnitControl ▸ UnitSelectControl

From now on, a `ref` prop will only be added _when_ it is wrapped by `forwardRef` or `contextConnect`.

> Do we need to make any changes to the contributing guidelines, especially re. handling refs as part of the TypeScript migration ?

Good point! I think we should remove the step about adding a `forwardRef`/`contextConnect` as part of the TS migration. They can be added when driven by an actual need, but shouldn't be part of the standard migration. Does that sound right? 5a63b29

File: packages/components/src/view/component.tsx
Comment (ciampo) [reply packages/components/src/view/component.tsx:37]: > In any case, these concerns are in follow-up territory.

Definitely, let's discuss in a follow-up!

**Review by ciampo — APPROVED** [review]
I believe this PR needs a rebase to include [this change](https://github.com/WordPress/gutenberg/pull/43736) to unblock e2e tests

> I think we should remove the step about adding a forwardRef/contextConnect as part of the TS migration. They can be added when driven by an actual need, but shouldn't be part of the standard migration. Does that sound right?

Sounds good to me 🆗 

> Good question. The good thing about the approach in this PR is that it immediately corrects these mistakes, with no refactoring required.

That sounds great, thank you for checking! As far as no components are expecting them to accept a `ref`, then we're good!

🚀

**Pull request merged**
